Move kube-dns ConfigMap creation/deletion out of federated services e2e tests.

madhusudancs opened the pull request:
These are now moved to federation-up.sh/federation-down.sh where the clusters are joined/unjoined. 

This PR depends on PR #41599. Please review only the last commit here. Other commits will be reviewed independently in their original PR.

Fixes: Issue #41589

```release-note
NONE
```

cc @kubernetes/sig-federation-pr-reviews @csbell

madhusudancs committed:
Move kube-dns ConfigMap creation/deletion out of federated services e2e tests to federation-up.sh/federation-down.sh where the clusters are joined/unjoined.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41601)
<!-- Reviewable:end -->

shashidharatd commented:
@madhusudancs, given few comments. PTAL

---

Reviewed 3 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[federation/cluster/federation-down.sh, line 28 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8ZRWOPpZM5G__-w5H:-Kd8ZRWOPpZM5G__-w5I:b1u17py) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L28)):*
> ```Shell
> 
> # join_cluster_to_federation joins the clusters in the local kubeconfig to federation. The clusters
> # and their kubeconfig entries in the local kubeconfig are created while deploying clusters, i.e. when kube-up is run.
> ```

Comment looks like copy-paste mistake, please correct.

---

*[federation/cluster/federation-down.sh, line 30 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8amlU012IxMD5cYu3:-Kd8amlU012IxMD5cYu4:b-q84qea) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L30)):*
> ```Shell
> # and their kubeconfig entries in the local kubeconfig are created while deploying clusters, i.e. when kube-up is run.
> function unjoin_clusters() {
>   "${host_kubectl}" config use-context "${FEDERATION_KUBE_CONTEXT}"
> ```

Did not get the purpose of setting the current context to FEDERATION_KUBE_CONTEXT.
I think we would be better off using --context flag wherever applicable? What do you think?

---

*[federation/cluster/federation-down.sh, line 32 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8_BFz1AcdD5z_INNH:-Kd8_BFz1AcdD5z_INNI:b-326wd4) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L32)):*
> ```Shell
>   "${host_kubectl}" config use-context "${FEDERATION_KUBE_CONTEXT}"
> 
>   local -r clusters=$("${host_kubectl}" -o jsonpath --template '{.items[*].metadata.name}')
> ```

This sentence looks like a mistake. it is not working i believe. Can you clarify please?

---

*[federation/cluster/federation-down.sh, line 40 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8_Ub6f1k8rl6zz_8J:-Kd8_Ub7aG8Os41VWk_i:b160j8l) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L40)):*
> ```Shell
>         --host-cluster-context="${HOST_CLUSTER_CONTEXT}"
> 
>     # Create kube-dns configmap in each cluster for kube-dns to accept
> ```

Create -- >Delete,
I think sentence needs change

---

*[federation/cluster/federation-down.sh, line 45 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8a3AgHNvsiHu3kW1v:-Kd8a3AgHNvsiHu3kW1w:bxttnv) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L45)):*
> ```Shell
>     # https://github.com/kubernetes/kubernetes/pull/39338.
>     # Remove this after the PR is merged.
>     "${host_kubectl}" delete configmap \
> ```

I think you intend to use --context="${cluster}", but missed it. am i right?
Otherwise, it would delete configmap in FEDERATION_KUBE_CONTEXT in a for loop, which is not desired.
Also the `host_kubectl` is set as below at the top.
host_kubectl="${KUBE_ROOT}/cluster/kubectl.sh --namespace=${FEDERATION_NAMESPACE}"

---

*[federation/cluster/federation-up.sh, line 118 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8dM_5fKvLuFCz12n5:-Kd8dM_6UyY51X0h_V4e:b-va0e1x) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-up.sh#L118)):*
> ```Shell
>     # Remove this after the PR is merged.
> 
>     "${KUBE_ROOT}/cluster/kubectl.sh" create configmap \
> ```

similar to above comment, need to use --context="${cluster}" to create the configmap in every cluster in for loop or create it once in federation context.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-:-Kd8egBY0S3-LV8gNQGR:b1x2o7l)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@shashidharatd sorry for this sloppy PR. I don't know what I was thinking, I just wanted to get this done and was in a rush.

---

Reviewed 2 of 4 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/cluster/federation-down.sh, line 28 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8ZRWOPpZM5G__-w5H:-Kd9ljbk5wIBYmjS62h0:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L28)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Comment looks like copy-paste mistake, please correct.
</blockquote></details>

Done.

---

*[federation/cluster/federation-down.sh, line 30 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8amlU012IxMD5cYu3:-Kd9lkqoWj8AED-sN0P9:b-3yr9v6) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L30)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Did not get the purpose of setting the current context to FEDERATION_KUBE_CONTEXT.
I think we would be better off using --context flag wherever applicable? What do you think?
</blockquote></details>

My plan was to just have this context as default for all the commands in the function. `kube-dns` ConfigMap deletion which is the only command that needs a different `--context` (where I accidentally missed specifying the context, **facepalm**) will go away once PR #39338 gets in.

But I see your point. If we are tearing down federation control plane anyway, what's the point in making it the default context. Made the necessary changes to explicitly use the context in all the commands.

---

*[federation/cluster/federation-down.sh, line 32 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8_BFz1AcdD5z_INNH:-Kd9ngH5YeJIgd8WQjHH:b-i6i4jp) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L32)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

This sentence looks like a mistake. it is not working i believe. Can you clarify please?
</blockquote></details>

I accidentally removed `get clusters` which changing `${KUBE_ROOT}/cluster/kubectl.sh` to `"${host_kubectl}"`. Fixed now. Thanks for catching that.

---

*[federation/cluster/federation-down.sh, line 40 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8_Ub6f1k8rl6zz_8J:-Kd9oFmU5ZHyPehc6ANC:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L40)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Create -- >Delete,
I think sentence needs change
</blockquote></details>

Done.

---

*[federation/cluster/federation-down.sh, line 45 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8a3AgHNvsiHu3kW1v:-Kd9oekOhzsIMJ0JMRI5:ba70oie) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-down.sh#L45)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

I think you intend to use --context="${cluster}", but missed it. am i right?
Otherwise, it would delete configmap in FEDERATION_KUBE_CONTEXT in a for loop, which is not desired.
Also the `host_kubectl` is set as below at the top.
host_kubectl="${KUBE_ROOT}/cluster/kubectl.sh --namespace=${FEDERATION_NAMESPACE}"
</blockquote></details>

You are right. Fixed.

---

*[federation/cluster/federation-up.sh, line 118 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-Kd8dM_5fKvLuFCz12n5:-KdKttltsOQrZGeQokj8:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea231cbbd9c0fe7ec0a100a729a6e8192ad9fc0d/federation/cluster/federation-up.sh#L118)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

similar to above comment, need to use --context="${cluster}" to create the configmap in every cluster in for loop or create it once in federation context.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-:-KdDG38ciBHEVhx_ec4i:borlvnn)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@shashidharatd PTAL now. It should be in a better shape, hopefully.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-:-KdKuADgunOonwJYn_nq:b-hotz0o)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@madhusudancs, the changes LGTM. Sorry for the delay

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41601#-:-KdM25tcVm0pZ96V6c0i:b-h5kiuh)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
/lgtm

madhusudancs commented:
Just rebased. Adding the LGTM and approvals back.

madhusudancs commented:
/lgtm
/approve

shashidharatd commented:
/lgtm

shashidharatd commented:
/lgtm cancel

shashidharatd commented:
/lgtm

mikedanese commented:
/approve

squash fixup commit?

madhusudancs commented:
@k8s-bot cvm gce e2e test this
@k8s-bot gci gce e2e test this

madhusudancs commented:
@k8s-oncall @saad-ali could you please manually merge this PR - https://github.com/kubernetes/kubernetes/pull/41601? Federation HEAD is completely broken right now - https://k8s-testgrid.appspot.com/google-federation and this PR is the fix.

madhusudancs commented:
@k8s-bot cvm gce e2e test this

madhusudancs commented:
@k8s-bot cvm gce e2e test this

madhusudancs commented:
@k8s-bot cross build this

saad-ali commented:
> @k8s-oncall @saad-ali could you please manually merge this PR - #41601? Federation HEAD is completely broken right now - https://k8s-testgrid.appspot.com/google-federation and this PR is the fix.

Manually merging to unblock federation.

```
16 successful checks
```

madhusudancs commented:
@saad-ali thanks!
